openapi3filter/req_resp_decoder.go

```diff
@@ -224,20 +224,62 @@ func decodeStyledParameter(param *openapi3.Parameter, input *RequestValidationIn
 
 func decodeValue(dec valueDecoder, param string, sm *openapi3.SerializationMethod, schema *openapi3.SchemaRef) (interface{}, error) {
 	var decodeFn func(param string, sm *openapi3.SerializationMethod, schema *openapi3.SchemaRef) (interface{}, error)
-	switch schema.Value.Type {
-	case "array":
-		decodeFn = func(param string, sm *openapi3.SerializationMethod, schema *openapi3.SchemaRef) (interface{}, error) {
-			return dec.DecodeArray(param, sm, schema)
+	if len(schema.Value.AnyOf) > 0 {
+		for _, sr := range schema.Value.AnyOf {
+			value, err := decodeValue(dec, param, sm, sr)
+			if value == nil && err == nil {
+				continue
+			}
+			return value, err
 		}
-	case "object":
-		decodeFn = func(param string, sm *openapi3.SerializationMethod, schema *openapi3.SchemaRef) (interface{}, error) {
-			return dec.DecodeObject(param, sm, schema)
+		return nil, nil
+	}
+	if len(schema.Value.AllOf) > 0 {
+		var value interface{}
+		var err error
+		for _, sr := range schema.Value.AllOf {
+			value, err = decodeValue(dec, param, sm, sr)
+			if value == nil || err != nil {
+				break
+			}
 		}
-	default:
-		decodeFn = dec.DecodePrimitive
+		return value, err
+	}
+	if len(schema.Value.OneOf) > 0 {
+		isMatched := 0
+		var value interface{}
+		for _, sr := range schema.Value.OneOf {
+			value, err := decodeValue(dec, param, sm, sr)
+			if err != nil {
+				return nil, err
+			}
+			if value != nil {
+				isMatched++
+			}
+		}
+		if isMatched == 1 {
+			return value, nil
+		}
+		return nil, nil
 	}
+	// TODO: if schema.Value.Not isn't nil , how to deal with it?
```

Comment:
As a user I'd prefer the decoder to fail on unhandled `not` rather than continue silently. Let's return an error when `schema.Value.Not` isn't `nil`.

Same for empty `schema.Value.Type`.

Reply:
I think if Not isn't nil, here we still can't return an error. because if Not isn't nil, it still doesn't mean an error i think . so i don't know how to deal with the same as an empty Value.Type?
Not means: according to json schema defination https://tools.ietf.org/html/draft-wright-json-schema-validation-00#section-5.25 

This keyword's value MUST be an object.  This object MUST be a valid
   JSON Schema.

   An instance is valid against this keyword if it fails to validate
   successfully against the schema defined by this keyword.

Reply:
Oh I mean we should error on non-nil Not because if it is set and we don't handle decoding Not then we should return an error, instead of the current behavior: silently continuing.

Reply:
Yes ,you are right. i have added this logic.